pydantic/types.py

```diff
@@ -148,12 +148,15 @@ def conbytes(*, strip_whitespace: bool = False, min_length: int = None, max_leng
 T = TypeVar('T')
 
 
-class ConstrainedList(List[T]):
-    __origin__ = list  # Needed for pydantic to detect that this is a list
+# This types superclass should be List[T], but cython chokes on that...
+class ConstrainedList(list):  # type: ignore
```

Comment:
I wonder if there is a way to put something into a conditional block (possibly even with `if TYPE_CHECKING`) so that we don't have to give up the generic parameter for mypy type-checking purposes. Or maybe it can still just be type-hinted as `List[T]`?

Reply:
Maybe something like

```py
if TYPE_CHECKING:
    class ConstrainedListSuperType(List[T]):
        pass
else:
    class ConstrainedListSuperType(list):
        pass

class ConstrainedList(ConstrainedListSuperType):
    ...
```

Reply:
I tried to implement this but I think I can only put the whole class under a conditional. The above does not work because for the typing to be correct you need: `class ConstrainedList(ConstrainedListSuperType[T]):` which gives the same issue...

This might be related to https://github.com/cython/cython/issues/2753

Reply:
I can put the whole block in `if TYPE_CHECKING:` but that duplicates quite a lot of code. So I would rather do it this way for now...